pkg/agent/reset.go

```diff
@@ -0,0 +1,223 @@
+	} else if sqlErr != nil {
+		return "", "", fmt.Errorf("querying hardware UUID and serial returned error: %w", sqlErr)
+	}
+
+	var resp []map[string]string
+	if err := json.Unmarshal(respBytes.Bytes(), &resp); err != nil {
+		return "", "", fmt.Errorf("unmarshalling response: %w", err)
+	}
+
+	if len(resp) < 1 {
+		return "", "", errors.New("no rows returned")
+	}
+
+	serial, ok := resp[0]["hardware_serial"]
+	if !ok {
+		return "", "", errors.New("hardware_serial missing from results")
+	}
+
+	hardwareUUID, ok := resp[0]["hardware_uuid"]
+	if !ok {
+		return "", "", errors.New("hardware_uuid missing from results")
+	}
+
+	return serial, hardwareUUID, nil
+}
+
+// valueChanged checks the knapsack for the given data and compares it to
+// currentValue. A value is considered changed only if the stored value was
+// previously set.
+func valueChanged(k types.Knapsack, currentValue string, dataKey []byte) bool {
+	storedValue, err := k.HostDataStore().Get(dataKey)
+	if err != nil {
+		k.Slogger().Error("could not get stored value", "err", err, "key", string(dataKey))
+		return false // assume no change
+	}
+
+	if storedValue == nil {
+		k.Slogger().Debug("value not previously stored, storing now", "key", string(dataKey))
+		if err := k.HostDataStore().Set(dataKey, []byte(currentValue)); err != nil {
+			k.Slogger().Error("could not store value", "err", err, "key", string(dataKey))
+		}
+		return false
+	}
+
+	if storedValue != nil && currentValue != string(storedValue) {
+		k.Slogger().Info("hardware-identifying value has changed", "key", string(dataKey))
+		return true
+	}
+
+	return false
+}
+
+// currentSecret retrieves the enrollment secret from either the knapsack or the filesystem,
+// depending on launcher configuration.
+func currentSecret(k types.Knapsack) (string, error) {
+	// Do we want to extract (and store) the tenant munemo, instead of storing the enrollment
+	// secret?
+	if k.EnrollSecret() != "" {
+		return k.EnrollSecret(), nil
+	}
+
+	if k.EnrollSecretPath() != "" {
+		content, err := os.ReadFile(k.EnrollSecretPath())
+		if err == nil {
+			return string(bytes.TrimSpace(content)), nil
+		}
+
+		return "", fmt.Errorf("could not read secret at enroll_secret_path %s: %w", k.EnrollSecretPath(), err)
+	}
+
+	return "", errors.New("enroll secret and secret path both unset")
+}
+
+// takeDatabaseBackup takes a backup of the current database and compresses it, storing
+// it in the root directory.
+func takeDatabaseBackup(k types.Knapsack) error {
+	backupFilepath := filepath.Join(k.RootDirectory(), "launcher.db.bak.zip")
+	f, err := os.Create(backupFilepath)
+	if err != nil {
+		return fmt.Errorf("creating backup database file: %w", err)
+	}
+	defer f.Close()
+
+	zipWriter := zip.NewWriter(f)
+	defer zipWriter.Close()
+
+	backupF, err := zipWriter.Create("launcher.db.bak")
+	if err != nil {
+		return fmt.Errorf("creating backup database inside zip: %w", err)
+	}
+
+	if err := k.BboltDB().View(func(tx *bbolt.Tx) error {
```

Comment:
This feels like a necessary-ish direct usage of BboltDB -- knapsack has abstracted bbolt use to per-store (i.e., per-bucket) operations only, and there isn't a nice way in the bbolt library to perform the backup per-bucket. There wasn't a clean way I could think of to update knapsack to accommodate this, and I'm not sure if it's worth the effort. Let me know if you have any ideas.